nerfstudio/fields/sdf_field.py

```diff
@@ -0,0 +1,449 @@
+"""
+
+from dataclasses import dataclass, field
+from typing import Optional, Type
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torch.nn.parameter import Parameter
+from torchtyping import TensorType
+from typing_extensions import Literal
+
+from nerfstudio.cameras.rays import RaySamples
+from nerfstudio.field_components.embedding import Embedding
+from nerfstudio.field_components.encodings import NeRFEncoding
+from nerfstudio.field_components.field_heads import FieldHeadNames
+from nerfstudio.field_components.spatial_distortions import SpatialDistortion
+from nerfstudio.fields.base_field import Field, FieldConfig
+
+try:
+    import tinycudann as tcnn
+except ImportError:
+    # tinycudann module doesn't exist
+    pass
+
+
+class SingleVarianceNetwork(nn.Module):
+    """Variance network in NeuS
+
+    Args:
+        init_val: initial value in NeuS variance network
+    """
+
+    def __init__(self, init_val):
+        super().__init__()
+        self.register_parameter("variance", nn.Parameter(init_val * torch.ones(1), requires_grad=True))
+
+    def forward(self, x: TensorType[1]) -> TensorType[1]:
+        """Returns current variance value"""
+        return torch.ones([len(x), 1], device=x.device) * torch.exp(self.variance * 10.0)
+
+    def get_variance(self) -> TensorType[1]:
+        """return current variance value"""
+        return torch.exp(self.variance * 10.0).clip(1e-6, 1e6)
+
+
+@dataclass
+class SDFFieldConfig(FieldConfig):
+    """SDF Field Config"""
+
+    _target: Type = field(default_factory=lambda: SDFField)
+    num_layers: int = 8
+    """Number of layers for geometric network"""
+    hidden_dim: int = 256
+    """Number of hidden dimension of geometric network"""
+    geo_feat_dim: int = 256
+    """Dimension of geometric feature"""
+    num_layers_color: int = 4
+    """Number of layers for color network"""
+    hidden_dim_color: int = 256
+    """Number of hidden dimension of color network"""
+    appearance_embedding_dim: int = 32
+    """Dimension of appearance embedding"""
+    use_appearance_embedding: bool = False
+    """Dimension of appearance embedding"""
+    bias: float = 0.8
+    """sphere size of geometric initializaion"""
+    geometric_init: bool = True
+    """Whether to use geometric initialization"""
+    inside_outside: bool = True
+    """whether to revert signed distance value, set to True for indoor scene"""
+    weight_norm: bool = True
+    """Whether to use weight norm for linear layer"""
+    use_grid_feature: bool = False
+    """Whether to use multi-resolution feature grids"""
+    divide_factor: float = 2.0
+    """Normalization factor for multi-resolution grids"""
+    beta_init: float = 0.1
+    """Init learnable beta value for transformation of sdf to density"""
+    encoding_type: Literal["hash", "periodic", "tensorf_vm"] = "hash"
+
+
+class SDFField(Field):
```

Comment:
`Field` has an abstract method `get_density`. How are you getting around that?

Reply:
Yea this was the guidance I was asking for in [this comment](https://github.com/nerfstudio-project/nerfstudio/pull/1444#issuecomment-1435783844). I could just add the method and make it do nothing (return None? pass?). Let me know your thoughts, this was originally there because the VolSDF method does use the get_density method but since that's not included in this PR I didn't think it made sense to bring in